### PR TITLE
Modeling issue 59 template script

### LINKED ISSUE #59: Improve modeling template script
- MLFlow logging on/off
- Speed up loading
- Clear logging outputs
  - Shorten GridSearch stopwords logging

### PR DESCRIPTION
**Pull request by lima-tango**
Done:
utils/modeling.py
- Posts class
- Training class
- MLFlowLogger class:
  - collects all params, metrics, and tags
  - logs to mlFlow on call of `log()`
  - saving of model to `./models`

modeling/model_template.py:
- template for modeling

ToDo:
- move `compute_and_log_metrics` to scoring and update `log_cm`
- where should `predict_with_threshold` live?
- dynamic creation of mlflow_params?
- caching of data loading

Closes #59

### CODE REVIEW

File: modeling/modeling.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+
+    Args:
+        y_true: The true target classification
+        y_pred: The predicted target classification
+        split: The split of the dataset ["test", "val", "train"]
+
+    Returns:
+        f1: The f1_score
+        precision: The precision_score
+        recall: The recall_score
```
Comment (chrsteck) [inline modeling/modeling.py:39]: I think it would be beneficial to print and return the confusion matrix

File: modeling/modeling.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+        X_train, y_train = self.data.get_X_y(split="train")
+        
+        logger.info(f"Fit model")
+        model = self.estimator.fit(X_train, y_train) #ToDo save to mlflow logger
+
+        # select best threshold if model implements predict_proba
+        if callable(getattr(model, "predict_proba", None)):
+            y_train_proba = model.predict_proba(X_train)[:, 1]
+            self.threshold = self.calculate_best_threshold(y_train, y_train_proba)
+            mlflow_params["threshold"] = self.threshold #ToDo save to mlflow logger
```
Comment (chrsteck) [inline modeling/modeling.py:172]: I  would prefer it, if the "best threshold" optimization would be optional so that you could turn it off or on.

**Review by chrsteck — COMMENTED** [review]
Two little comments. The rest looks great!

File: modeling/modeling.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+        # store best parameters if model is GridSearch
+        if isinstance(model, GridSearchCV):
+            best_params = model.best_params_
+            if 'vectorizer__stop_words' in best_params.keys() and best_params['vectorizer__stop_words']!=None:
+                best_params['vectorizer__stop_words'] = "NLTK-German"
+            mlflow_params["best_params"] = best_params #ToDo save to mlflow logger
+
+        self.model = model
+        #self.save_model_to_drive(model)
+        return model
```
Comment (dominikmn) [inline modeling/modeling.py:183]: With 
```
model = self.estimator.fit(X_train, y_train)
```
`model` is just the same as `self.estimator`
I guess you want something like
```
model_best = self.estimator.best_estimator_
``` 
after the fitting?

Relevant text from the [sklearn documentation](https://scikit-learn.org/stable/developers/develop.html#fitting) 

> The method should return the object (self). This pattern is useful to be able to implement quick one liners in an IPython session such as:

File: modeling/modeling.py (hunk 1 above)
Comment (lima-tango) [reply modeling/modeling.py:39]: done (but needs to be integrated with utils.scoring)

File: modeling/modeling.py (hunk 3 above)
Comment (lima-tango) [reply modeling/modeling.py:183]: done

File: modeling/modeling.py (hunk 2 above)
Comment (lima-tango) [reply modeling/modeling.py]: done

File: utils/modeling.py [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+    f1 = f1_score(y_true, y_pred)
+    precision = precision_score(y_true, y_pred)
+    recall = recall_score(y_true, y_pred)
+
+    tn, fp, fn, tp = confusion_matrix(y_true, y_pred).ravel()
+    cm = {'TN':tn, 'FP':fp, 'FN':fn, 'TP':tp}
+
+    logger.info(f"Performance on {split} set: F1 = {f1:.1f}, precision = {precision:.1%}, recall = {recall:.1%}")
+    logger.info(f"Confusion matrix: {cm}")
+    return f1, precision, recall, cm
```
Comment (lima-tango) [inline utils/modeling.py:52]: I will move this function to utils.scoring if you agree.

File: utils/modeling.py [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+            
+
+def predict_with_threshold(y_pred_proba: np.array, threshold: float) -> np.array:
+    return (y_pred_proba >= threshold).astype(int)
```
Comment (lima-tango) [inline utils/modeling.py:56]: In which utils do we want to put this function?

File: utils/modeling.py [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+            label: The label to use as target. Default is all available labels
+            split: The data-split to use. One of ['train', 'val', 'test']
+        
+        Returns:
+            X: The feature (text column) of the posts
+            y: The target annotations
+        """
+        if split:
+            filter_frame = pd.read_csv(f'./data/ann2_{split}.csv', header=None, index_col=0, names=['id_post'])
+            df = self.df.merge(filter_frame, how='inner', on='id_post')
```
Comment (lima-tango) [inline utils/modeling.py:82]: Do we want to cache?

File: modeling/model_template.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    # MLFlow params have limited characters, therefore stopwords must not be given as list
+    grid_search_params = param_grid.copy()
+    grid_search_params["vectorizer__stop_words"] = ["NLTK-German", None]
+
+    mlflow_params = {
+        "vectorizer": "count",
+        "normalization": "lower",
+        "model": "NaiveBayes",
+        "grid_search_params": grid_search_params,
+    }
```
Comment (lima-tango) [inline modeling/model_template.py:49]: I tried to do this dynamically, but in my opinion it doesn't work well and is not easily readable.

File: modeling/model_template.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        "cycle2": True,
+    }
+
+    TARGET_LABELS = ['label_argumentsused', 'label_discriminating', 'label_inappropriate',
+        'label_offtopic', 'label_personalstories', 'label_possiblyfeedback',
+        'label_sentimentnegative', 'label_sentimentpositive',]
+    
+    IS_DEVELOPMENT = True
+
+    data = m.Posts()
```
Comment (chrsteck) [inline modeling/model_template.py:60]: How can we integrate the data augmentation in this step? This is a crucial step imho.

**Review by chrsteck — CHANGES_REQUESTED** [review]
If we can integrate the augmentation function in the model template, ignore my comment. The rest looks great 👍

**Pull request merged**